core/client/src/main/java/alluxio/client/file/FileSystemUtils.java

```diff
@@ -157,15 +159,26 @@ public static long persistFile(FileSystem fs, AlluxioURI uri, URIStatus status)
       FileInStream in = closer.register(fs.openFile(uri, options));
       AlluxioURI dstPath = new AlluxioURI(status.getUfsPath());
       UnderFileSystem ufs = UnderFileSystem.get(dstPath.toString());
-      String parentPath = dstPath.getParent().toString();
-      if (!ufs.exists(parentPath)) {
-        URIStatus parentStatus = fs.getStatus(uri.getParent());
-        Permission parentPerm = new Permission(parentStatus.getOwner(), parentStatus.getGroup(),
-            (short) parentStatus.getMode());
-        MkdirsOptions parentMkdirsOptions = new MkdirsOptions().setCreateParent(true)
-            .setPermission(parentPerm);
-        if (!ufs.mkdirs(parentPath, parentMkdirsOptions)) {
-          throw new IOException("Failed to create " + parentPath);
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = uri.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      while (!ufs.exists(curUfsPath.toString())) {
+        URIStatus curDirStatus = fs.getStatus(curAlluxioPath);
+        Permission perm = new Permission(curDirStatus.getOwner(), curDirStatus.getGroup(),
+            (short) curDirStatus.getMode());
+        ufsDirsToMakeWithOptions.push(new Pair<>(curUfsPath.toString(),
+            new MkdirsOptions().setCreateParent(false).setPermission(perm)));
+
+        curAlluxioPath = curAlluxioPath.getParent();
+        curUfsPath = curUfsPath.getParent();
+      }
+      while (!ufsDirsToMakeWithOptions.empty()) {
+        Pair<String, MkdirsOptions> ufsDirAndPerm = ufsDirsToMakeWithOptions.pop();
+        if (!ufs.mkdirs(ufsDirAndPerm.getFirst(), ufsDirAndPerm.getSecond())) {
+          throw new IOException("Failed to create " + ufsDirAndPerm.getFirst() + " with permission "
+              + ufsDirAndPerm.getSecond().toString());
```

Comment:
Not really related to this PR, would it be easier to call something on the master which would do the right thing? Ideally client should not directly talk to ufs.